### PR TITLE
changes for EINVAL issue reported by customer, latest version changes

### PR DESCRIPTION
**Pull request by mohitpavan**
1. https://github.com/Azure/load-testing/issues/95 there is an issue with some EINVAL where we use some ExecFile command and there is a breaking change in node18,20 and it started failing in windows.
ref : https://github.com/nodejs/node/issues/52554
2. we missed updating the API-version to latest preview version, done now.
3. added the report in the loadTest output folder.
4. fixed the undefined issues when there are no statistics.
5. updated the pr build checks to check both the windows and ubuntu.
6. there is a bug, when the test run is polled for completion, the TR is deleted, then it goes to the infinite loop, changed it to proper one and included the errors in the same.
7. added the check for the file checks which the pipeline produces in the E2E of pr build.
8. in time of validation of test files, if the deletion of test occurs, then there is error that it shows validation succeeded and then fails while creating the TR.

### CODE REVIEW

File: .github/workflows/pr_check_load_test-windows.yml [hunk 1]
```diff
@@ -13,18 +13,22 @@ on:
       - main
 
 env:
-  LOAD_TEST_RESOURCE: "rbac-test"
+  LOAD_TEST_RESOURCE: "mohit-1"
   LOAD_TEST_RESOURCE_GROUP: "radhika-test"
   COMMIT_ID : ${{ github.sha }}
 
+permissions:
+  id-token: write # This is required for requesting the JWTcontents: read  # This is required for actions/checkout
```
Comment (radhikajain99) [inline .github/workflows/pr_check_load_test-windows.yml:21]: nit: Correct the comment. This - 
contents: read # This is required for actions/checkout

Should be in next line

File: lib/mappers.js [hunk 2]
```diff
@@ -54,12 +53,12 @@ var configFiles = [];
 var zipFiles = [];
 var token = "";
 var resourceId = "";
-var subscriptionID = "";
+var subscriptionID = "7c71b563-0dc0-4bc0-bcf6-06f8f0516c7a";
```
Comment (radhikajain99) [inline lib/mappers.js:56]: Revert?

File: lib/mappers.js [hunk 3]
```diff
@@ -54,12 +53,12 @@ var configFiles = [];
...
 var environment = "AzureCloud";
 var armTokenScope = "https://management.core.windows.net";
 var dataPlaneTokenScope = "https://loadtest.azure-dev.com";
 var armEndpoint = "https://management.azure.com";
-var tenantId = "";
+var tenantId = "72f988bf-86f1-41af-91ab-2d7cd011db47";
```
Comment (radhikajain99) [inline lib/mappers.js:61]: Same here

File: sample-test-jmx-windows.yaml [hunk 4]
```diff
@@ -0,0 +1,19 @@
+version: v0.1
```
Comment (radhikajain99) [inline sample-test-jmx-windows.yaml:1]: Can we use same files for ubuntu and windows? We can parameterize the testName to avoid conflicts

File: sample-test-url-windows.yaml [hunk 5]
```diff
@@ -0,0 +1,5 @@
+version: v0.1
```
Comment (radhikajain99) [inline sample-test-url-windows.yaml:1]: Same comment here

File: src/mappers.ts [hunk 6]
```diff
@@ -21,12 +20,12 @@ var configFiles: string[] = [];
 var zipFiles : string[]=[];
 var token = "";
 var resourceId = "";
-var subscriptionID = "";
+var subscriptionID = "7c71b563-0dc0-4bc0-bcf6-06f8f0516c7a";
```
Comment (radhikajain99) [inline src/mappers.ts:23]: Same, revert?

File: src/util.ts [hunk 7]
```diff
@@ -33,8 +33,9 @@ const validConditionList = {
     'error': ['>']
 }
 export module apiConstants {
-    export const tm20240301previewVersion = '2024-03-01-preview';
-    export const tm2023Version = '2023-04-01-preview';
+    export const latestVersion = '2024-05-01-preview';
+    // export const tm20240301PreviewVersion = '2024-03-01-preview';
```
Comment (radhikajain99) [inline src/util.ts:37]: nit: Remove instead of commenting?

File: .github/workflows/pr_check_load_test-windows.yml (hunk 1 above)
Comment (mohitpavan) [reply .github/workflows/pr_check_load_test-windows.yml]: done

File: lib/mappers.js (hunk 2 above)
Comment (mohitpavan) [reply lib/mappers.js]: by mistake, done now

File: lib/mappers.js (hunk 3 above)
Comment (mohitpavan) [reply lib/mappers.js]: done

File: sample-test-jmx-windows.yaml (hunk 4 above)
Comment (mohitpavan) [reply sample-test-jmx-windows.yaml]: changed

File: src/util.ts (hunk 7 above)
Comment (mohitpavan) [reply src/util.ts]: removed

File: src/mappers.ts (hunk 6 above)
Comment (mohitpavan) [reply src/mappers.ts]: done

File: .github/workflows/pr_check_load_test-ubuntu.yml [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+      - releases/*
+      - main
+
+env:
+  LOAD_TEST_RESOURCE: "mohit-1"
+  LOAD_TEST_RESOURCE_GROUP: "radhika-test"
+  COMMIT_ID : ${{ github.sha }}
+
+permissions:
+  id-token: write # This is required for requesting the JWTcontents: read  # This is required for actions/checkout
```
[the remark this answers is not in the document]
Comment (mohitpavan) [reply .github/workflows/pr_check_load_test-ubuntu.yml]: done

File: .github/workflows/pr_check_load_test.yml [hunk 9]
```diff
@@ -43,13 +47,35 @@ jobs:
...
+            echo "::set-output name=GUID::$(powershell -Command "[guid]::NewGuid().ToString()")"
+          else
+            echo "::set-output name=GUID::$(uuidgen)"
+          fi
+      shell : bash
+
+    - name: Set up testName
+      shell: bash
+      run: |
+        # replace 'originalTestName' with the original test name in the file
```
Comment (radhikajain99) [inline .github/workflows/pr_check_load_test.yml:69]: nit: 1st comment is not needed now?

Comment (mohitpavan) [reply .github/workflows/pr_check_load_test.yml]: done

File: src/util.ts [hunk 10]
```diff
@@ -105,7 +104,7 @@ export function checkFileType(filePath: string, fileExtToValidate: string): bool
 }
 export async function printTestDuration(vusers:string, startTime:Date, endTime : Date, testStatus : string) 
 {
-    console.log("TestRun completed\n");
+    console.log("Summary generation completed\n");
```
Comment (radhikajain99) [inline src/util.ts:107]: Why is this changed?

**Review by radhikajain99 — APPROVED** [review]

File: src/util.ts (hunk 10 above)
Comment (mohitpavan) [reply src/util.ts:107]: this is because i added the testRun completed log somewhere else, and this is actually printed when the summary, results and report also gets generated only.

File: .github/workflows/pr_check_load_test.yml [hunk 11]
```diff
@@ -13,18 +13,22 @@ on:
       - main
 
 env:
-  LOAD_TEST_RESOURCE: "rbac-test"
+  LOAD_TEST_RESOURCE: "mohit-1"
```
Comment (mitsha-microsoft) [inline .github/workflows/pr_check_load_test.yml:16]: Can we use some other name? or get it from repo variables?

File: .github/workflows/pr_check_load_test.yml [hunk 12]
```diff
@@ -43,13 +47,49 @@ jobs:
...
+
+    - name: Generate GUID
+      id: guid
+      run: |
+          if [[ "${{ runner.os }}" == "Windows" ]]; then
+            echo "::set-output name=GUID::$(powershell -Command "[guid]::NewGuid().ToString()")"
+          else
+            echo "::set-output name=GUID::$(uuidgen)"
+          fi
+      shell : bash
```
Comment (mitsha-microsoft) [inline .github/workflows/pr_check_load_test.yml:64]: shell should not be required?

File: .github/workflows/pr_check_load_test.yml [hunk 13]
```diff
@@ -43,13 +47,49 @@ jobs:
...
+
+    - name: Set up testName
+      shell: bash
+      run: |
+        # replacing 'testId'(the thing present in the file.) with the new test name ('${{ steps.guid.outputs.GUID }}') in the file.
```
Comment (mitsha-microsoft) [inline .github/workflows/pr_check_load_test.yml:69]: nit: `with the new testId`

File: .github/workflows/pr_check_load_test.yml [hunk 14]
```diff
@@ -43,13 +47,49 @@ jobs:
...
+        # copying the files to a new file and editing the testId for each file in the copied file.
+        cp ${{ matrix.configFile }} test-config-${{ steps.guid.outputs.GUID }}.yaml
+        sed -i 's/testId/${{ steps.guid.outputs.GUID }}/g' test-config-${{ steps.guid.outputs.GUID }}.yaml
```
Comment (mitsha-microsoft) [inline .github/workflows/pr_check_load_test.yml:72]: sed is not present in windows machines?

Comment (mohitpavan) [reply .github/workflows/pr_check_load_test.yml:72]: yeah sed is not present in windows

File: .github/workflows/pr_check_load_test.yml (hunk 12 above)
Comment (mohitpavan) [reply .github/workflows/pr_check_load_test.yml:64]: if [[ "${{ runner.os }}" == "Windows" ]]; then
            echo "::set-output name=GUID::$(powershell -Command "[guid]::NewGuid().ToString()")"
          else
            echo "::set-output name=GUID::$(uuidgen)"
          fi
          
          this is bash script, and windows wont understand, if i dont define shell

Comment (mohitpavan) [reply .github/workflows/pr_check_load_test.yml:64]: windows mein default is power shell and this is not power shell script

File: .github/workflows/pr_check_load_test.yml (hunk 11 above)
Comment (mohitpavan) [reply .github/workflows/pr_check_load_test.yml:16]: yes, please update the secrets or env in the repo

File: .github/workflows/pr_check_load_test.yml (hunk 13 above)
Comment (mohitpavan) [reply .github/workflows/pr_check_load_test.yml]: done

File: .github/workflows/pr_check_load_test.yml (hunk 11 above)
Comment (mohitpavan) [reply .github/workflows/pr_check_load_test.yml]: done

**Review by mitsha-microsoft — DISMISSED** [review]

File: src/main.ts [hunk 15]
```diff
@@ -329,22 +335,25 @@ async function createTestRun() {
...
 async function getTestRunAPI(testRunId:string, testStatus:string, startTime:Date) 
 {   
-    var urlSuffix = "test-runs/"+testRunId+"?api-version="+util.apiConstants.tm20240301previewVersion;
+    var urlSuffix = "test-runs/"+testRunId+"?api-version="+util.apiConstants.latestVersion;
     urlSuffix = baseURL+urlSuffix;
     while(!util.isTerminalTestStatus(testStatus)) 
     {
         let header = await map.getTestRunHeader();
         let testRunResult = await util.httpClientRetries(urlSuffix,header,'get',3,"");
         let testRunObj:any = await util.getResultObj(testRunResult);
```
Comment (radhikajain99) [inline src/main.ts:344]: Same here?

File: src/mappers.ts [hunk 16]
```diff
@@ -325,7 +324,7 @@ export async function getSubName() {
   } catch (err: any) {
     const message =
       `An error occurred while getting credentials from ` +
-      `Azure CLI: ${err.stack}`;
+      `Azure CLI: ${err.message}` + ` for getting subscription name.`;
```
Comment (radhikajain99) [inline src/mappers.ts:327]: `for getting subscription name` should be added before err.message?

File: src/mappers.ts [hunk 17]
```diff
@@ -344,7 +343,7 @@ async function getAccessToken(aud: string) {
   } catch (err: any) {
     const message =
       `An error occurred while getting credentials from ` +
-      `Azure CLI: ${err.stack}`;
+      `Azure CLI: ${err.message}` + ` for getting access token.`;
```
Comment (radhikajain99) [inline src/mappers.ts:346]: Same here

**Review by radhikajain99 — APPROVED** [review]

**Comment by cmyllis** [comment]
@mohitpavan Any new release containing the fix?
